Add automatic image rotation based on EXIF data.

ezvirtual opened the pull request:
This was previously supported in ATContentTypes and a discussion in p.a.contenttypes lead to the suggestion that this was the best place to re-implement the functionality.

[Plone.app.contenttypes issue with discussion on the subject](https://github.com/plone/plone.app.contenttypes/issues/32#issuecomment-66964387)
[@maartenkling's post on adding this functionality for dexterity images] (https://www.fourdigits.nl/blog/auto-rotate-dexterity-images-based-on-exif-info)

I originally added a subscriber for Archetypes images however the current code needs changing to use Archetypes accessor methods.

I'm relatively not quite sure on the correct protocol around this process - this is my first contribution to plone core - I'm more than happy to receive any hints / feedback!! :smiley:

pbauer commented:
jenkins: http://jenkins.plone.org/job/pull-request-5.0/320/

jensens commented:
@ezvirtual I cant find you in the list of Plone contributors. If you already signed the agreement please send an email to agreements@plone.org to fix this, other wise please become a contributor, information at: https://plone.org/foundation/contributors-agreement

ezvirtual commented:
@jensens I just forwarded copy of my original contributor agreement through to the supplied email address. Cheers.

maartenkling commented:
Nice!

jensens commented:
@ezvirtual thx, sorry, i found you after double checking. for some reason plone does not show up here https://github.com/ezvirtual :(

jensens commented:
since this is more complex code, may you please add a test for it?

polyester commented:
The Contributor Agreement for @ezvirtual is there (thanks, Ezra!) and is a member of the Plone org. So that's triple-checked now ;-)

pbauer commented:
@ezvirtual there are some test-failures: http://jenkins.plone.org/job/pull-request-5.0/320/console
```AttributeError: StringIO instance has no attribute 'data'```

mauritsvanrees commented:
I see this pull request and I see the same or similar changes on the 2.0.x branch, without the pull request. Do we want this on the 2.0.x branch, so on Plone 4.3?

pbauer commented:
@ezvirtual Thanks for the work! Some suggestions: I want the subscriber enabled for all dextierty content (not only p.a.c). The subscriber should iterate over all fields (in schema and behaviors). If a field is a image-field first check if there actually is any data stored and then try get exif and rotate the image based on that data.

vangheem commented on plone/namedfile/image_exif_rotation.py:
```diff
@@ -0,0 +1,123 @@
+
+        self.image.data = image2.getvalue()
+
+        self.reindexObject()
```
Why is reindexObject done here?

vangheem commented on plone/namedfile/image_exif_rotation.py:
```diff
@@ -0,0 +1,123 @@
+            'Rotated 90 CCW': 8}
+
+
+def exif_rotation(self, event):
```
Style here. Please don't use "self". "self" is meant to be used for class instance methods. These are functions and variable names should be "obj" or "content". This needs to be changed in all the functions.

vangheem commented:
Well, sorry, I missed that tests were already running for this. Here is another test run for good measure I guess: http://jenkins.plone.org/job/pull-request-5.0/323/

vangheem commented on plone/namedfile/image_exif_rotation.py:
```diff
@@ -0,0 +1,123 @@
+        img = img.transpose(method)
+        img.save(image2, fmt, quality=88)
+
+        self.image.data = image2.getvalue()
```
I think you might want to do image._setData(image2.getvalue()) instead here.

ezvirtual commented:
Thanks for those pointers everyone!

@mauritsvanrees - I reverted my (overly keen / accidental) push on the 2.0.x branch and made an equivalent pull request there. A side question - is it worth adding something to the README making it explicit / really obvious that the 2.0.x branch is used by Plone 4.3? 

I have implemented @vangheem's suggestions for both branches:  5f1efe9522fd552c5bce588baa0b6cfc766b870e 4f79d8e7e9a98f9633b279a43c255b443e77d083

@pbauer: Thanks for running the tests, can you (/anyone) point me in the right direction to be able to work on resolving the couple of errors? I guess the first step is to get the tests running locally? Is there any reason why the [test report doesn't report any failures whilst the console does have errors](http://jenkins.plone.org/job/pull-request-5.0/323/testReport/)?

Enabling the subscriber for all dexterity content types makes sense to me. I didn't (/don't) have enough accumulated knowledge to know if this interface is found on all dexterity images or not - plone.namedfile.interfaces.IImageScaleTraversable I can look at getting the subscriber to interate through fields / behaviours and next, followed by @jensens suggestion of writing some tests (I probably need slightly more time on my hands than I have this week!)

My assumption is that I:
- Make my changes
- close this pull request
- create a new pull request

Is that vaguely sensible or is there a better way of proceeding?

Thanks to all for the encouragement and patience to answer many questions... In the course of writing this comment I see that there is a lot of [excellent documentation] (http://docs.plone.org/develop/plone-coredev/intro.html) that I ~~need to~~ should have already read!!

pbauer commented:
@ezvirtual I think first you have to write your own test. One should add a default ct "image" with a image-file that needs to be rotated. The other should register a custom content-type with image-fields where you should do the same. 
That the testrunner did not collect the errors in the report is weird but probably not related. 
The method to rotate and mirror should be written in such a way that it can be reused to "manually" do these actions. I'll also add some notes to the code itself.

pbauer commented on plone/namedfile/image_exif_rotation.py:
```diff
@@ -0,0 +1,121 @@
+    if obj.portal_type == 'Image':
+        if obj.image:
+
+            mirror, rotation = get_exif_orientation(obj)
```
Don't pass the object, only the image (maybe as StringIO?)

pbauer commented on plone/namedfile/image_exif_rotation.py:
```diff
@@ -0,0 +1,121 @@
+            if rotation:
+                transform = AUTO_ROTATE_MAP.get(rotation, None)
+                if transform is not None:
+                    transform_image(obj, transform)
```
Again don't pass the obj. Set the new image on the field after transforming.

pbauer commented on plone/namedfile/image_exif_rotation.py:
```diff
@@ -0,0 +1,121 @@
+
+def exif_rotation(obj, event):
+    if obj.portal_type == 'Image':
+        if obj.image:
```
iterate over fields:
```python
from plone.dexterity.utils import iterSchemata
from plone.namedfile.interfaces import INamedBlobImageField

for schema in iterSchemata(obj):
    for fieldname, field in schema.namesAndDescriptions():
        if INamedBlobImageField.providedBy(field):
```

loechel commented:
The Idea looks good, but please don't do it like that. 

If we want to have it on all Images of plone.namedfile, hock it into https://github.com/plone/plone.namedfile/blob/master/plone/namedfile/file.py so that you just work on the data and not on the souraounding contenttype. 

The Exifdata sould be handled by getImageInfo() and the transformed data should be written.

So no subscriber is needed and all Images will be handled this way.

ezvirtual commented:
Hi @loechel, that sounds like a sensible way to handle this. I'll look at changing my current implementation to this when I have a bit more time on my hands. Cheers

loechel commented:
@ezvirtual if you are at ploneconf I would help you during a sprint to make it finalized.
